Lib/unittest/test/test_discovery.py

```diff
@@ -406,10 +406,34 @@ def _find_tests(start_dir, pattern, namespace=None):
         top_level_dir = os.path.abspath('/foo/bar')
         start_dir = os.path.abspath('/foo/bar/baz')
         self.assertEqual(suite, "['tests']")
-        self.assertEqual(loader._top_level_dir, top_level_dir)
+        self.assertEqual(loader._top_level_dir, os.path.abspath('/foo'))
         self.assertEqual(_find_tests_args, [(start_dir, 'pattern')])
         self.assertIn(top_level_dir, sys.path)
 
+    def test_discover_should_not_persist_top_level_dir_between_calls(self):
+        original_isfile = os.path.isfile
+        original_isdir = os.path.isdir
+        original_sys_path = sys.path[:]
+        def restore():
+            os.path.isfile = original_isfile
+            os.path.isdir = original_isdir
+            sys.path[:] = original_sys_path
+        self.addCleanup(restore)
+
+        os.path.isfile = lambda path: True
```

Comment:
Is `unittest.mock.patch` a better option here instead of `addCleanup`?

```python
def test_discover_should_not_persist_top_level_dir_between_calls(self):
    with unittest.mock.patch('os.path.isfile', return_value=True):
        with unittest.mock.patch('os.path.isdir', return_value=True):
            loader = unittest.TestLoader()
            loader.suiteClass = str
            dir = '/foo/bar'
            top_level_dir = '/foo'

            loader.discover(dir, top_level_dir=top_level_dir)
            self.assertEqual(loader._top_level_dir, None)

            loader._top_level_dir = dir2 = '/previous/dir'
            loader.discover(dir, top_level_dir=top_level_dir)
            self.assertEqual(loader._top_level_dir, dir2)
```

Reply:
Nowadays that can be `self.enterContext`. But, as is, the PR is consistent with other nearby tests; if this needs changing it would be best to change all of them (in a separate PR).